pkg/controllers/workgenerator/controller.go

```diff
@@ -515,13 +544,30 @@ func (r *Reconciler) syncAllWork(ctx context.Context, resourceBinding *fleetv1be
 	return true, updateAny.Load(), nil
 }
 
+// syncApplyStrategy syncs the apply strategy specified on a ClusterResourceBinding object
+// to a Work object.
+func (r *Reconciler) syncApplyStrategy(
+	ctx context.Context,
+	resourceBinding *fleetv1beta1.ClusterResourceBinding,
+	existingWork *fleetv1beta1.Work,
+) (bool, error) {
+	// Skip the update if no change on apply strategy is needed.
+	if equality.Semantic.DeepEqual(existingWork.Spec.ApplyStrategy, resourceBinding.Spec.ApplyStrategy) {
+		return false, nil
+	}
+
+	// Update the apply strategy on the work.
+	existingWork.Spec.ApplyStrategy = resourceBinding.Spec.ApplyStrategy.DeepCopy()
+	if err := r.Client.Update(ctx, existingWork); err != nil {
+		klog.ErrorS(err, "Failed to update the apply strategy on the work", "work", klog.KObj(existingWork), "binding", klog.KObj(resourceBinding))
+		return true, controller.NewUpdateIgnoreConflictError(err)
+	}
+	klog.V(2).InfoS("Successfully updated the apply strategy on the work", "work", klog.KObj(existingWork), "binding", klog.KObj(resourceBinding))
+	return true, nil
+}
+
 // areAllWorkSynced checks if all the works are synced with the resource binding.
 func areAllWorkSynced(existingWorks map[string]*fleetv1beta1.Work, resourceBinding *fleetv1beta1.ClusterResourceBinding, _, _ string) bool {
-	syncedCondition := resourceBinding.GetCondition(string(fleetv1beta1.ResourceBindingWorkSynchronized))
```

Comment:
Hi Ryan! For this removal I am wondering if I could check with you to see if there's anything I missed/overlooked.

Reply:
There seems to be a bug with the current code: say, if there is a binding with a missing master resource snapshot, and the user issues an apply strategy change, the ideal outcome would be that the binding (and its works) receives the new apply strategy and continues to report the latest Applied/Available condition (under the stale resource snapshot). However, with this sync cond check, Fleet will always deem the work(s) to be out of sync (as the sync cond becomes stale with the apply strategy change), and the user will see an OverridenFailed error (cond) even if there are no overrides specified at all.

Reply:
And, another thing is, if the sync cond is True (and up to date), the works are guaranteed to be synchronized, so the check below seems redundant I assume? Instead the only case where we could skip the check below seems to be verifying if the sync cond is False and up to date, which would imply that things have already gone wrong.

Reply:
Apologies if I miss or misunderstand anything here 😣

Reply:
> as the sync cond becomes stale with the apply strategy change

I think the problem is exactly this. The out of the loop update of the work before all the other logic probably breaks more than this part of the code in some corner cases. I think the original goal was to "isolate" this logic but it might have backfired.

Reply:
Changes made after the latest round of discussed changes.